pkg/volume/util/operationexecutor/operation_executor.go

```diff
@@ -387,6 +391,82 @@ func (oe *operationExecutor) DetachVolume(
 	return oe.pendingOperations.Run(
 		volumeToDetach.VolumeName, "" /* podName */, detachFunc)
 }
+func (oe *operationExecutor) BulkVerifyVolumes(
+	attachedVolumes map[types.NodeName][]AttachedVolume,
+	actualStateOfWorld ActualStateOfWorldAttacherUpdater) {
+
+	// A map of plugin names and nodes on which they exist with volumes they manage
+	bulkVerifyPluginsByNode := make(map[string]map[types.NodeName][]*volume.Spec)
+	volumeSpecMapByPlugin := make(map[string]map[*volume.Spec]v1.UniqueVolumeName)
+
+	for node, nodeAttachedVolumes := range attachedVolumes {
+		for _, volumeAttached := range nodeAttachedVolumes {
+			volumePlugin, err :=
+				oe.operationGenerator.GetVolumePluginMgr().FindPluginBySpec(volumeAttached.VolumeSpec)
+
+			if err != nil || volumePlugin == nil {
+				glog.Errorf(
+					"VolumesAreAttached.FindPluginBySpec failed for volume %q (spec.Name: %q) on node %q with error: %v",
+					volumeAttached.VolumeName,
+					volumeAttached.VolumeSpec.Name(),
+					volumeAttached.NodeName,
+					err)
+				continue
+			}
+
+			pluginName := volumePlugin.GetPluginName()
+
+			if volumePlugin.SupportsBulkVolumeVerification() {
+				pluginNodes, pluginNodesExist := bulkVerifyPluginsByNode[pluginName]
+
+				if !pluginNodesExist {
+					pluginNodes = make(map[types.NodeName][]*volume.Spec)
+				}
+
+				volumeSpecList, nodeExists := pluginNodes[node]
+				if !nodeExists {
+					volumeSpecList = []*volume.Spec{}
+				}
+				volumeSpecList = append(volumeSpecList, volumeAttached.VolumeSpec)
+				pluginNodes[node] = volumeSpecList
+
+				bulkVerifyPluginsByNode[pluginName] = pluginNodes
+				volumeSpecMap, mapExists := volumeSpecMapByPlugin[pluginName]
+
+				if !mapExists {
+					volumeSpecMap = make(map[*volume.Spec]v1.UniqueVolumeName)
+				}
+				volumeSpecMap[volumeAttached.VolumeSpec] = volumeAttached.VolumeName
+				volumeSpecMapByPlugin[pluginName] = volumeSpecMap
+				continue
+			}
+
+			// If node doesn't support Bulk volume polling it is best to poll individually
+			nodeError := oe.VerifyVolumesAreAttached(nodeAttachedVolumes, node, actualStateOfWorld)
```

Comment:
This function is much larger than other functions in this file. The logic is a little bit tricky, likethe "break" is very critical in line 449. Could we change something like following?
bulkVerifyVolumesByPlugin := make(map[string]map[types.NodeName][]AttachedVolume)
nonBulkVerifyVolumesByPlugin := make(map[string]map[types.NodeName][]AttachedVolume)
for node, nodeAttachedVolumes := range attachedVolumes {
 		for _, volumeAttached := range nodeAttachedVolumes {
 			// get volumePlugin
                        // If plugin support bulkVolume, put in bulkVerifyVolumesByPlugin, 
                        // otherwise put in nonBulkVerifyVolumesByPlugin

 for pluginName, bulkVerifyVolumes := range bulkVerifyVolumesByPlugin
             // oe.operationtor.GenerateBulkVolumeBerifyFunc( bulkVerifyVolumes, actualStateOfWorld)

for pluginName, nonBulkVerifyVolumes := range nonBulkVerifyVolumesByPlugin
           for nodeName, volumes := range nonBulkVerifyVolumes 
           //oe.VerifyVolumesAttached(volumes, nodeName, actualStateOfWorld)

This might involves more looping, but I feel the logic is easier and code change is smaller.